python/tests/test_session.py

```diff
@@ -255,3 +255,13 @@ def test_border_cases():
                 "e0": "twitter_property_e_0#header_row=true",
             }
         )
+
+
+def test_with():
+    with graphscope.session(cluster_type="hosts") as sess:
+        assert graphscope.get_default_session() == sess
+
+    sess = graphscope.session(cluster_type="hosts")
+    with sess:
+        pass
+    assert sess.info["status"] == "closed"
```

Comment:
This test case, as well as the `self.close()` in `__exit__` looks strange.

What's the behavior of `with <established session>` in other frameworks, e.g., tensorflow?

Reply:
It doesn't close session.

Reply:
`sess.info["status"] == "closed"`?

Reply:
My understanding is that our primary goal using `with` is to automatically release resources, is that correct?

Reply:
Any problem with the `sess.info["status"] == "closed"`? Could you make it more clear 😂

Reply:
With `with expression` statement the `__enter__` and `__exit__` should be symmetry, i.e., `open(....)` in `__enter__` and `close(...)` in `__exit__`. The `expression` it self shouldn't handle such stuffs.

If we `.as_default()` in `__enter__` then I think the expected behavior in `__exit__` should be _pop_ the default session, rather than closing it.

Reply:
But the open stage is done in `__init__`, the way you are saying is to solely do the register default and deregister stuff.

> If we .as_default() in __enter__ then I think the expected behavior in __exit__ should be pop the default session, rather than closing it.
Yes, I actually closed it after poped the default session.

Reply:
So the problem becomes, should we close the session in `__exit__`? What is already known is we don't open the session in `__enter__`.